src/scripts/tbtcv2-rewards/utils.ts

```diff
@@ -0,0 +1,504 @@
+      }
+
+      instances.push(instanceTimestampsVersionInfo)
+
+      const dataInstance = instancesData.get(instance.metric.instance)
+      lastRegisteredVersion.set(
+        instanceTimestampsVersionInfo.buildVersion,
+        instanceTimestampsVersionInfo.lastRegisteredTimestamp
+      )
+      if (dataInstance !== undefined) {
+        dataInstance.version = Object.fromEntries(lastRegisteredVersion)
+      } else {
+        // Should not happen
+        console.error("Instance must be present for a given rewards interval.")
+      }
+    }
+
+    // Sort instances in ascending order by first registration timestamp
+    instances.sort((a, b) =>
+      a.firstRegisteredTimestamp > b.firstRegisteredTimestamp ? 1 : -1
+    )
+
+    return instances
+  }
+
+  public convertToObject(map: Map<string, InstanceParams>) {
+    let obj: { [k: string]: any } = {}
+    map.forEach((value: InstanceParams, key: string) => {
+      obj[key] = value
+    })
+
+    return obj
+  }
+
+  public async queryPrometheus(url: string, params: any): Promise<any> {
+    try {
+      const { data } = await axios.get(url, { params: params })
+
+      return data
+    } catch (error) {
+      if (axios.isAxiosError(error)) {
+        console.log("error message: ", error.message)
+        return error.message
+      } else {
+        console.log("unexpected error: ", error)
+        return "An unexpected error occurred"
+      }
+    }
+  }
+
+  public async getBootstrapData(
+    startRewardsTimestamp: number,
+    endRewardsTimestamp: number
+  ) {
+    // Query for bootstrap data that has peer instances grouped by operators
+    const queryBootstrapData = `${this.prometheusAPI}/query_range`
+    const paramsBootstrapData = {
+      query: `sum by(chain_address)({job='${this.prometheusJob}'})`,
+      start: startRewardsTimestamp,
+      end: endRewardsTimestamp,
+      step: OPERATORS_SEARCH_QUERY_STEP,
+    }
+
+    const bootstrapData = await this.queryPrometheus(
+      queryBootstrapData,
+      paramsBootstrapData
+    )
+    return bootstrapData.data.result
+  }
+
+  public async isVersionSatisfied(
+    operatorAddress: string,
+    rewardsInterval: number,
+    startRewardsTimestamp: number,
+    endRewardsTimestamp: number,
+    clientReleases: any,
+    instancesData: Map<string, InstanceParams>
+  ) {
+    // keep-core client already has at least 2 released versions
+    const latestClient = clientReleases[0].split("_")
+    const latestClientTag = latestClient[0]
+    const latestClientTagTimestamp = Number(latestClient[1])
+    const secondToLatestClient = clientReleases[1].split("_")
+    const secondToLatestClientTag = secondToLatestClient[0]
+
+    const instances = await this.processBuildVersions(
+      operatorAddress,
+      rewardsInterval,
+      instancesData
+    )
+
+    let isVersionSatisfied = true
+    const upgradeCutoffDate = latestClientTagTimestamp + ALLOWED_UPGRADE_DELAY
+    if (upgradeCutoffDate < startRewardsTimestamp) {
```

Comment:
Please c&p the latest code around versioning validation from `main`. We modified some code recently.